bundle/config/mutator/python/override.go

```diff
@@ -0,0 +1,313 @@
+package python
+
+import (
+	"fmt"
+
+	"github.com/databricks/cli/libs/dyn"
+	"github.com/databricks/cli/libs/dyn/merge"
+)
+
+// overrideMode restricts which changes are allowed during the override process
+type overrideMode string
+
+const OverrideModeAppend overrideMode = "append"
+const OverrideModeAppendOrModify overrideMode = "append-or-update"
+
+// isModified keeps track of whether a value has been modified during the override process
+type isModified struct {
+	value bool
+}
+
+var Modified = isModified{value: true}
+var NotModified = isModified{value: false}
+
+type overrideFunc func(dyn.Value, dyn.Value) (dyn.Value, isModified, error)
+
+// OverrideByPath overrides bundle config 'leftRoot' with 'rightRoot' values in 'path', keeping
+// 'location' if values haven't changed. This allows to keep track of where the values
+// were last modified for error reporting.
+//
+// 'path' must point to a mapping or not exist, otherwise, error is returned
+//
+// 'mode' allows to restrict what kinds of changes are allowed in 'path':
+// - append: only new values can be added to mapping
+// - append-or-modify: new values can be added, or existing values can be modified
+func OverrideByPath(leftRoot dyn.Value, rightRoot dyn.Value, path dyn.Path, mode overrideMode) (dyn.Value, error) {
+	_, leftErr := dyn.GetByPath(leftRoot, path)
+	_, rightErr := dyn.GetByPath(rightRoot, path)
+
+	if leftErr != nil && rightErr != nil {
+		return leftRoot, nil
+	}
+
+	leftRoot, err := insertMappingIfAbsent(leftRoot, path)
+
+	if err != nil {
+		return dyn.InvalidValue, err
+	}
+
+	rightRoot, err = insertMappingIfAbsent(rightRoot, path)
+
+	if err != nil {
+		return dyn.InvalidValue, err
+	}
+
+	return dyn.Map(leftRoot, path.String(), func(p dyn.Path, left dyn.Value) (dyn.Value, error) {
+		leftMapping := left.MustMap()
+		rightMapping, err := dyn.GetByPath(rightRoot, p)
+
+		if err != nil {
+			return dyn.InvalidValue, err
+		}
+
+		out, _, err := overrideMapping(
```

Comment:
My read on the override function argument is that you added it such that you can check the override mode and produce an error if something was modified when this isn't allowed. Is this correct?

If so, I think a neater approach would be to not have this override function argument and always recurse into the override function, and always check the mode on modifications. The mode could be stored in a struct that you make these override functions a member of.

Reply:
I need to think about it. This function works differently depending on the depth where we call it. Because we want to handle appends and modifications differently. Append in nested object becomes modification on a level above. One option to fix it is to return a tree of what has been changed, then we don't need to put this logic into "override" function.

Reply:
I don't think we need to return a complex object.

When we call this function for a map, the following can happen:
* Existing entries have been updated
* New entries have been added
* A combination of both

As long as we can capture these in a status object, the caller can decide what to do.